scene/3d/skeleton_modifier_3d.cpp

```diff
@@ -0,0 +1,158 @@
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "skeleton_modifier_3d.h"
+
+void SkeletonModifier3D::_validate_property(PropertyInfo &p_property) const {
+	//
+}
+
+PackedStringArray SkeletonModifier3D::get_configuration_warnings() const {
+	PackedStringArray warnings = Node3D::get_configuration_warnings();
+	if (skeleton_id.is_null()) {
+		warnings.push_back(RTR("Skeleton3D node not set! SkeletonModifier3D must be child of Skeleton3D or set a path to an external skeleton."));
+	}
+	return warnings;
+}
+
+void SkeletonModifier3D::rebind() {
+	_update_skeleton();
+	_rebind();
+}
+
+void SkeletonModifier3D::_rebind() {
+	//
+}
+
+/* Skeleton3D */
+
+Skeleton3D *SkeletonModifier3D::get_skeleton() const {
+	return Object::cast_to<Skeleton3D>(ObjectDB::get_instance(skeleton_id));
+}
+
+void SkeletonModifier3D::_update_skeleton_path() {
+	skeleton_id = ObjectID();
+
+	// Make sure parent is a Skeleton3D.
+	Skeleton3D *sk = Object::cast_to<Skeleton3D>(get_parent());
+	if (sk) {
+		skeleton_id = sk->get_instance_id();
+	}
+}
+
+void SkeletonModifier3D::_update_skeleton() {
+	if (!is_inside_tree()) {
+		return;
+	}
+	Skeleton3D *old_sk = get_skeleton();
+	_update_skeleton_path();
+	Skeleton3D *new_sk = get_skeleton();
+	if (old_sk != new_sk) {
+		_skeleton_changed(old_sk, new_sk);
+	}
+	update_configuration_warnings();
+}
+
+void SkeletonModifier3D::_skeleton_changed(Skeleton3D *p_old, Skeleton3D *p_new) {
+	//
+}
```

Comment:
`_skeleton_changed` is missing GDVIRTUAL.

Perhaps a signal would make more sense than a virtual function?
I do not quite understand why we can't use NOTIFICATION_PARENTED and NOTIFICATION_ENTER_TREE

Reply:
Nice pick. This was a holdover from when Modifier had an external_skeleton, and there were multiple unnecessary methods associated with it, now I removed them. However, I don't think it is worth making this SIGNAL, so I will leave it as is.